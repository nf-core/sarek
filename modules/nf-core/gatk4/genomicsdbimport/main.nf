process GATK4_GENOMICSDBIMPORT {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/gatk4:4.5.0.0--py36hdfd78af_0':
        'biocontainers/gatk4:4.5.0.0--py36hdfd78af_0' }"

    input:
    tuple val(meta), path(vcf), path(tbi), path(interval_file), val(interval_value), path(wspace)
    val   run_intlist
    val   run_updatewspace
    val   input_map

    output:
    tuple val(meta), path("$prefix")        , optional:true, emit: genomicsdb
    tuple val(meta), path("$updated_db")    , optional:true, emit: updatedb
    tuple val(meta), path("*.interval_list"), optional:true, emit: intervallist
    path "versions.yml"                                    , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args   ?: ''
    prefix   = task.ext.prefix ?: "${meta.id}"

    // settings for running default create gendb mode
    input_command = input_map ? "--sample-name-map ${vcf[0]}" : vcf.collect(){"--variant $it"}.join(' ')

    genomicsdb_command = "--genomicsdb-workspace-path ${prefix}"
    interval_command = interval_file ? "--intervals ${interval_file}" : "--intervals ${interval_value}"
    updated_db = ""

    // settings changed for running get intervals list mode if run_intlist is true
    if (run_intlist) {
        genomicsdb_command = "--genomicsdb-update-workspace-path ${wspace}"
        interval_command = "--output-interval-list-to-file ${prefix}.interval_list"
    }

    // settings changed for running update gendb mode. input_command same as default, update_db forces module to emit the updated gendb
    if (run_updatewspace) {
        genomicsdb_command = "--genomicsdb-update-workspace-path ${wspace}"
        interval_command = ''
        updated_db = "${wspace}"
    }

    def avail_mem = 3072
    if (!task.memory) {
        log.info '[GATK GenomicsDBImport] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }
    """
    gatk --java-options "-Xmx${avail_mem}M -XX:-UsePerfData" \\
        GenomicsDBImport \\
        $input_command \\
        $genomicsdb_command \\
        $interval_command \\
        --tmp-dir . \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    prefix   = task.ext.prefix ?: "${meta.id}"

    genomicsdb_command = "--genomicsdb-workspace-path ${prefix}"
    interval_command = interval_file ? "--intervals ${interval_file}" : "--intervals ${interval_value}"
    updated_db = ""

    // settings changed for running get intervals list mode if run_intlist is true
    if (run_intlist) {
        genomicsdb_command = "--genomicsdb-update-workspace-path ${wspace}"
        interval_command = "--output-interval-list-to-file ${prefix}.interval_list"
    }

    // settings changed for running update gendb mode. input_command same as default, update_db forces module to emit the updated gendb
    if (run_updatewspace) {
        genomicsdb_command = "--genomicsdb-update-workspace-path ${wspace}"
        interval_command = ''
        updated_db = "${wspace}"
    }

    def stub_genomicsdb = genomicsdb_command == "--genomicsdb-workspace-path ${prefix}" ? "touch ${prefix}" : ""
    def stub_interval   = interval_command == "--output-interval-list-to-file ${prefix}.interval_list" ? "touch ${prefix}.interval_list" : ""
    def stub_update     = updated_db != "" ? "touch ${wspace}" : ""

    """
    ${stub_genomicsdb}
    ${stub_interval}
    ${stub_update}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """
}
