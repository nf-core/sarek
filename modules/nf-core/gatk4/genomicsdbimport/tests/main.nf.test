nextflow_process {

    name "Test Process GATK4_GENOMICSDBIMPORT"
    script "../main.nf"
    process "GATK4_GENOMICSDBIMPORT"

    tag "modules"
    tag "modules_nfcore"
    tag "untar"
    tag "gatk4"
    tag "gatk4/genomicsdbimport"

    test("test_gatk4_genomicsdbimport_create_genomicsdb") {

        when {
            process {
                """
                // [meta, vcf, tbi, interval, interval_value, workspace ]
                input[0] = [ [ id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf.gz', checkIfExists: true) ,
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf.gz.tbi', checkIfExists: true) ,
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.interval_list', checkIfExists: true) ,
                    [] ,
                    [] ]
                // run_intlist
                input[1] = false
                // run_updatewspace
                input[2] = false
                // input_map
                input[3] = false

                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                //{ assert snapshot(file(process.out.updatedb.get(0).get(1)).list().sort()).match() }
                //{ assert snapshot(process.out.intervallist.get(0).get(1)).match() }
                { assert snapshot(
                    file(process.out.genomicsdb.get(0).get(1)).list().sort(),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("test_gatk4_genomicsdbimport_get_intervalslist") {

        setup {
            run("UNTAR") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = Channel.of([ [], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/test_genomicsdb.tar.gz', checkIfExists: true) ])
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = Channel.of([ [id:"test"], [], [], [], []]).combine(UNTAR.out.untar.map{ it[1] })
                // run_intlist
                input[1] = true
                // run_updatewspace
                input[2] = false
                // input_map
                input[3] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                //{ assert snapshot(file(process.out.genomicsdb.get(0).get(1)).list().sort()).match() }
                //{ assert snapshot(file(process.out.updatedb.get(0).get(1)).list().sort()).match() }
                { assert snapshot(
                    process.out.intervallist.get(0).get(1),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("test_gatk4_genomicsdbimport_update_genomicsdb") {

        setup {
            run("UNTAR") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = Channel.of([ [], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/test_genomicsdb.tar.gz', checkIfExists: true) ])
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = Channel.of([ [id:"test"], file( params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test2.genome.vcf.gz' , checkIfExists: true), file( params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test2.genome.vcf.gz.tbi' , checkIfExists: true), [], []]).combine(UNTAR.out.untar.map{ it[1] })
                // run_intlist
                input[1] = false
                // run_updatewspace
                input[2] = true
                // input_map
                input[3] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                //{ assert snapshot(file(process.out.genomicsdb.get(0).get(1)).list().sort()).match() }
                //{ assert snapshot(process.out.intervallist.get(0).get(1)).match() }
                { assert snapshot(
                    file(process.out.updatedb.get(0).get(1)).list().sort(),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("test_gatk4_genomicsdbimport_stub") {

        options "-stub"

        when {
            process {
                """
                // [meta, vcf, tbi, interval, interval_value, workspace ]
                input[0] = [ [ id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf.gz', checkIfExists: true) ,
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf.gz.tbi', checkIfExists: true) ,
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.interval_list', checkIfExists: true) ,
                    [] ,
                    [] ]
                // run_intlist
                input[1] = false
                // run_updatewspace
                input[2] = false
                // input_map
                input[3] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match()}
            )
        }
    }
}
