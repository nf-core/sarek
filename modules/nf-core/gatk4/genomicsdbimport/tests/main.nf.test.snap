{
    "test_gatk4_genomicsdbimport_get_intervalslist": {
        "content": [
            "test.interval_list:md5,4c85812ac15fc1cd29711a851d23c0bf",
            [
                "versions.yml:md5,c1233a04213021aa66599a36e0fb28cc"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-09T10:42:51.836379"
    },
    "test_gatk4_genomicsdbimport_create_genomicsdb": {
        "content": [
            [
                "__tiledb_workspace.tdb",
                "callset.json",
                "chr22$1$40001",
                "vcfheader.vcf",
                "vidmap.json"
            ],
            [
                "versions.yml:md5,c1233a04213021aa66599a36e0fb28cc"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-09T10:42:36.846239"
    },
    "test_gatk4_genomicsdbimport_update_genomicsdb": {
        "content": [
            [
                "__tiledb_workspace.tdb",
                "callset.json",
                "chr22$1$40001",
                "vcfheader.vcf",
                "vidmap.json"
            ],
            [
                "versions.yml:md5,c1233a04213021aa66599a36e0fb28cc"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-09T10:43:09.00769"
    },
    "test_gatk4_genomicsdbimport_stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,c1233a04213021aa66599a36e0fb28cc"
                ],
                "genomicsdb": [
                    [
                        {
                            "id": "test"
                        },
                        "test:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "intervallist": [
                    
                ],
                "updatedb": [
                    
                ],
                "versions": [
                    "versions.yml:md5,c1233a04213021aa66599a36e0fb28cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-09T10:43:20.921712"
    }
}